dashboard/lib/services/globally_unique_identifiers.rb

```diff
@@ -0,0 +1,75 @@
+module Services
+  # A module to assist with providing globally-unique identifiers for models
+  # whose unique identifiers are only unique within the context of a given
+  # relationship.
+  module GloballyUniqueIdentifiers
+    # Simple helper which builds out a regex from the given models. Requires
+    # that the models define a KEY_CHAR_RE which can be used to identify valid
+    # `key` values for the model.
+    #
+    # Specifically, given Bar and Baz, we will get a regex which matches
+    # strings of the form `bar_key/baz_key`
+    def self.build_key_re(models)
+      keys = models.map do |model|
+        "(?<#{model.name.demodulize}>#{model::KEY_CHAR_RE}+)"
+      end
+      return Regexp.new(keys.join('/'))
+    end
+
+    # Returns a key which can be used by to globally and uniquely identify the
+    # given Vocabulary object
+    #
+    # @see GloballyUniqueIdentifiers.find_vocab
+    def self.build_vocab_key(vocab)
+      return unless vocab&.course_version&.course_offering.present?
```

Comment:
it'll get `nil`, and so will try to query or serialize with `nil` as a value

Reply:
I am not super familiar with the crowdin codepaths -- this is fine with me if you feel like you have a clear picture of why a nil value is safe to return here.

Reply:
oh, to be clear: a vocabulary (or resource) object _not_ having associated course version and offering values is in fact an invalid state for vocabulary or resource data to be in. It's just the case that generally in the I18N sync we want to ignore invalid data rather than raising an error.

I'll add some logging here, though!

Reply:
that makes sense -- the piece I'm missing is that returning nil here means the i18n sync will safely ignore it, as opposed to that nil value accidentally getting used as a key in some output file, for example.